server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -714,4 +777,38 @@ private void maybeFinish() {
     private static RemoteTransportException wrapRemoteClusterFailure(String clusterAlias, Exception e) {
         return new RemoteTransportException("error while communicating with remote cluster [" + clusterAlias + "]", e);
     }
+
+    static Map<String, OriginalIndices> getIndicesFromSearchContexts(Map<ShardId, SearchContextIdForNode> searchContexts,
+                                                                     IndicesOptions indicesOptions) {
+        final Map<String, List<String>> indices = new HashMap<>();
+        for (Map.Entry<ShardId, SearchContextIdForNode> entry : searchContexts.entrySet()) {
+            String clusterAlias = entry.getValue().getClusterAlias() == null ?
+                RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY : entry.getValue().getClusterAlias();
+            indices.computeIfAbsent(clusterAlias, k -> new ArrayList<>()).add(entry.getKey().getIndexName());
+        }
+        return indices.entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> new OriginalIndices(e.getValue().toArray(String[]::new), indicesOptions)));
+    }
+
+    static List<SearchShardIterator> getSearchShardsFromSearchContexts(ClusterState clusterState, OriginalIndices originalIndices,
+                                                                       String localClusterAlias,
+                                                                       Map<ShardId, SearchContextIdForNode> searchContexts,
+                                                                       TimeValue keepAlive) {
+        final List<SearchShardIterator> iterators = new ArrayList<>(searchContexts.size());
+        for (Map.Entry<ShardId, SearchContextIdForNode> entry : searchContexts.entrySet()) {
+            final ShardId shardId = entry.getKey();
+            final ShardIterator shards = OperationRouting.getShards(clusterState, shardId);
+            final List<ShardRouting> matchingNodeFirstRoutings = new ArrayList<>();
+            for (ShardRouting shard : shards) {
+                if (shard.currentNodeId().equals(entry.getValue().getNode())) {
+                    matchingNodeFirstRoutings.add(0, shard);
+                } else {
+                    matchingNodeFirstRoutings.add(shard);
```

Comment:
why do we use shards from other nodes?  Is it that they will not have matching searchContext any way?

Reply:
We discussed this offline. We are adding every shard copy to the iterator for these two reasons:

- Reuse the existing error handling that we have in TransportSearchAction
- We can plugin the retry logic if the target shards are not available